scipy/signal/signaltools.py

```diff
@@ -428,7 +455,177 @@ def fftconvolve(in1, in2, mode="full"):
+    else:
+        raise ValueError('mode is invalid')
+
+    # see whether the Fourier transform convolution method or the direct
+    # convolution method is faster (discussed in scikit-image PR #1792)
+    direct_time = (x.size * h.size * _prod(out_shape))
+    fft_time = sum_builtin(n * np.log(n) for n in (x.shape + h.shape +
+                                               tuple(out_shape)))
+    return big_O_constant * fft_time < direct_time
+
+
+def _reverse_and_conj(x):
+    """
+    Reverse array `x` in all dimensions and perform the complex conjugate
+    """
+    reverse = [slice(None, None, -1)] * x.ndim
+    return x[reverse].conj()
+
+
+def _np_conv_ok(volume, kernel, mode):
+    """
+    See if numpy supports convolution of x and y (if both are 1D ndarrays and
+    of the appropriate shape).  Numpy's 'same' mode uses the size of the
+    larger input, not the first.
+    """
+    np_conv_ok = volume.ndim == kernel.ndim == 1
+    return np_conv_ok and (volume.size >= kernel.size or mode != 'same')
+
+
+def choose_conv_method(in1, in2, mode='full', method='auto', try_=False):
+    """
+    A method to find the fastest convolution method.
+
+    Parameters
+    ----------
+    in1 : array_like
+        The first argument passed into the convolution function.
+    in2 : array_like
+        The second argument passed into the convolution function.
+    mode : str {'full', 'valid', 'same'}, optional
+        A string indicating the size of the output:
+
+        ``full``
+           The output is the full discrete linear convolution
+           of the inputs. (Default)
+        ``valid``
+           The output consists only of those elements that do not
+           rely on the zero-padding.
+        ``same``
+           The output is the same size as `in1`, centered
+           with respect to the 'full' output.
+    method : str {'fft', 'auto', 'direct'}, optional
+        A str indicating which convolution method would like to be used.
+    try_ : bool, optional
+        If True, run and time the convolution with `in1` and `in2` under both
+        methods and return the fastest method. If False (default), predict the
+        fastest method using precomputed values.
+
+    Returns
+    -------
+    method : str
+        A string indicating which convolution method is fastest, either
+        'direct' or 'fft'
+
+    Raises
+    ------
+    ValueError
+        If the given convolution method does not support the input.
+
+    See also
+    --------
+    convolve
+
+    Notes
+    -----
+    For large n, this function is accurate and can decide upon the fastest
+    method to perform the convolution.  However, this function is not as
+    accurate for small n (when any dimension in the input or output is small).
+
+    """
+    volume = asarray(in1)
+    kernel = asarray(in2)
+
+    if try_:
+        times = {}
+        for method in ['fft', 'direct']:
+            start = time.time()
```

Comment:
this should probably use the timeit module instead.  I have a function here that does it https://gist.github.com/endolith/2d35f3ea8f64cd2d02d1b49b43f25a21#file-timeit_utils-py-L32